### PR TITLE
Implement users tokens

### PR DESCRIPTION
**Pull request by Abyss777**
Here is nice addition to users disabling.
Main reason is to give link user to traccar site without username/password

User can have a token, that can be used to authorization via GET parameter instead of username/password:
`https://demo.traccar.org/api/session?token=s5XpfXKZ3zLi7X4A`

Only Admin can add/change token to user.

Also hardcoded minimal token length and symbols to use in token.

Security note: tokens should be used very carefully and mostly with read only users in combination with expiring. Only over HTTPS. Because `GET` parameters is absolutely unsecured and cached/logged everywhere.

### CODE REVIEW

File: src/org/traccar/api/resource/SessionResource.java [hunk 1]
```diff
@@ -69,6 +69,12 @@ public User get() throws SQLException {
                     userId = user.getId();
                     request.getSession().setAttribute(USER_ID_KEY, userId);
                 }
+            } else if (request.getParameter("token") != null) {
```
Comment (tananaev) [inline src/org/traccar/api/resource/SessionResource.java]: Request parameter can be declared as a method argument. See position resource for example.

File: src/org/traccar/api/resource/UserResource.java [hunk 2]
```diff
@@ -66,7 +66,8 @@ public Response update(@PathParam("id") long id, User entity) throws SQLExceptio
                 || old.getAdmin() != entity.getAdmin()
                 || old.getReadonly() != entity.getReadonly()
                 || old.getDisabled() != entity.getDisabled()
-                || old.getDeviceLimit() != entity.getDeviceLimit()) {
+                || old.getDeviceLimit() != entity.getDeviceLimit()
+                || !old.getToken().equals(entity.getToken())) {
```
Comment (tananaev) [inline src/org/traccar/api/resource/UserResource.java]: What if it's null? Is it possible?

File: src/org/traccar/database/PermissionsManager.java [hunk 3]
```diff
@@ -39,6 +40,7 @@
     private volatile Server server;
 
     private final Map<Long, User> users = new ConcurrentHashMap<>();
+    private final Map<Long, String> usersTokens = new HashMap<>();
```
Comment (tananaev) [inline src/org/traccar/database/PermissionsManager.java]: I think you should flip the parameters. Lookup by token seems to be more frequent operation than creating new token.

File: src/org/traccar/model/User.java [hunk 4]
```diff
@@ -180,6 +180,31 @@ public void setDeviceLimit(int deviceLimit) {
...
 
+    private String token;
+
+    public String getToken() {
+        return token;
+    }
+
+    public void setToken(String token) {
+        if (token != null && !token.isEmpty()) {
+            if (validateToken(token)) {
```
Comment (tananaev) [inline src/org/traccar/model/User.java]: I suggest to keep validation in the setter for now. Something like this:
```java
if (token != null && !token.isEmpty()) {
    if (!token.matches("^[a-zA-Z0-9]{16,}$")) {
        throw new IllegalArgumentException("Illegal token");
    }
    this.token = token;
}
```

**Comment by Abyss777** [comment]
It's up to you.  I'm not sure I'll fix all comments before release.

**Comment by Abyss777** [comment]
Sorry for missed null checks and `@QueryParam`.
`userId`-`token` map looked more simple to update for me.
Also combined validation to setter as you suggested.

**Comment by tananaev** [comment]
I have a couple of small comments, but I'll fix them myself to completely this before release.

**Pull request merged**